client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -674,12 +684,30 @@ Response get() throws IOException {
                     e.addSuppressed(exception);
                     throw e;
                 }
-                //try and leave the exception untouched as much as possible but we don't want to just add throws Exception clause everywhere
+                /*
+                 * Wrap and rethrow whatever exception we received, copying the type
+                 * where possible so the synchronous API looks as much as possible
+                 * like the asynchronous API. We wrap the exception so that the caller's
+                 * signature shows up in any exception we throw.
+                 */
+                if (exception instanceof ResponseException) {
```

Comment:
I would have thought to use reflection here to look for a constructor of the appropriate shape (taking an `Exception` of the same concrete type), falling back to `String` and calling `Throwable#initCause` if not, and then falling back to `RuntimeException`. What do you think of that?

Reply:
I thought about the reflection solution here and rejected it, mostly because we tend to dislike it in the rest of Elasticsearch. I think the trouble with reflection here is that it is difficult to reason about. Not in the "what is going to happen?" sense, but in the "are the ctors that I call going to do the right thing with the things I give them?" sense.

For what it is worth I'd like to abstract async http client from the Elasticsearch client eventually and I don't think the reflection based approach would play well with that.

I certainly understand that this will change the exceptions that we throw in the cases where I don't have an explicit branch though.

Reply:
Or is it a way to wrap the exceptions thrown (all or parts) already in the async call into a "CustomException" ? kind of like the `ElasticsearchException` ? This might make exception a little bit redundant but it keeps the consistence of the 2 sides of sync and async API. (If the exception parse back stuff in sync listener is too tricky to do) ?

Reply:
I think the exception unwrapping is going to be tricky to get perfect. I'd be ok with some kind of rethrown exception. I mean, ultimately, that is what we end up with if we fall through all the `instanceof` blocks. We could save ourselves the trouble and wrap the same way every time. I'm honestly not sure what'd be easier for callers though.

@javanna, what do you think? You've been working in this area a lot longer than I have.

Reply:
As far as I understand what it all boils down to here is the different IOExceptions that can be thrown by the underlying http client, there may be more than the ones that we have a branch for. 

I think the current solution is good enough, I even wonder if anybody is ever going to catch socket timeout rather connect timeout etc. maybe all those could just be generic IOExceptions like we already do below and one has to look at the cause to see what it really is? I think that would be reasonable too, but we do want to rethrow ResponseException as a proper ResponseException as it's our own (like we already do).

I do not have anything against the proposed reflection solution either, but maybe what we have now is easier to reason about and we can always add branches if we missed something.

Reply:
Maybe we should document what we do with exceptions in the sync methods, so people know what to expect.

Reply:
another idea, not sure though how much it makes sense, could be to add our own specific runtime exception (instead of a generic one) with a good name that indicates that it's used only for wrapping, and always use that one, and document well that users have to look at the cause. Then we may want to also remove all the `throws IOException` from the `performRequest` methods. Not sure though, maybe this is the silliest way of addressing this problem, it is hard to evaluate how much this would change things for users, I suspect not much but not sure.

Reply:
I keep coming back to "I'd like to make async httpclient an implementation detail one day" and I think this kind of `if instanceof then rethrow` is the kind of logic I'm going to need to do that one day. So, I think I'll stick with this approach for entirely selfish reasons. As a side effect, I think it should *mostly* be transparent to users. Except for branches I haven't caught....

++ on adding javadoc for this.

Reply:
@nik9000 do you think it would be helpful to add another branch for [SSLException](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLException.html) ?

Reply:
> @nik9000 do you think it would be helpful to add another branch for SSLException ?

Fine with me! I can do it.